From 089b145652bb63b377d7c1a618b9e9da2e2b231c Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Wed, 24 Apr 2024 08:29:10 +0300
Subject: [PATCH] [FSW-4077] fs1: efuse: restructure code

This patch restructures efuse code to avoid
code duplication across multiple platforms. All
device specific includes are removed from common
code to make it scalable in future.

Change-Id: If4692b5c616fbcf01e536149a888230dbdca692e
---

diff --git a/arch/fs1/arc/common.c b/arch/fs1/arc/common.c
index 5e74a1a..3f77631 100644
--- a/arch/fs1/arc/common.c
+++ b/arch/fs1/arc/common.c
@@ -146,7 +146,18 @@
 	}
 }
 
-static void boot_rom_set_efuse_read(void)
+enum boot_rom_status boot_rom_read_config_fuse(void)
+{
+	if ((boot_rom_fuse_read(IO_EFUSE_SYSCFG_OFFSET, &boot_rom_global.config_fuse)
+	    != BOOT_ROM_SUCCESS) ||
+	    (boot_rom_fuse_read(IO_EFUSE_SYSCFG2_OFFSET, &boot_rom_global.config2_fuse)
+	    != BOOT_ROM_SUCCESS))
+		return BOOT_ROM_FAIL;
+
+	return BOOT_ROM_SUCCESS;
+}
+
+void boot_rom_set_efuse_read(void)
 {
 	struct fs1_cold_rst_data cold_rst_data;
 
@@ -159,27 +170,3 @@
 	regmap_writel(cold_rst_data.data, IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_COLD_RST_REG_2));
 }
 
-void boot_rom_fuse(void)
-{
-	if (boot_rom_fuse_init() != BOOT_ROM_SUCCESS)
-		goto efuse_fail_halt;
-
-	if (boot_rom_fuse_read(IO_EFUSE_SYSCFG_OFFSET, &boot_rom_global.config_fuse) != BOOT_ROM_SUCCESS)
-		goto efuse_fail_halt;
-
-	if(boot_rom_fuse_read(IO_EFUSE_SYSCFG2_OFFSET, &boot_rom_global.config2_fuse) != BOOT_ROM_SUCCESS)
-		goto efuse_fail_halt;
-
-	if (!fuse_valid_volt_prot_pattern(boot_rom_global.config_fuse))
-		goto efuse_fail_halt;
-
-	boot_rom_set_efuse_read();
-	/* TODO: FS1 SKU decession pending */
-
-	/* TODO: Enable FAKE FUSE & RMA support */
-	return;
-
-efuse_fail_halt:
-	boot_rom_status_update(PROGRESS_INIT_EFUSE_INIT_FAIL);
-	boot_rom_halt();
-}
diff --git a/arch/fs1/arc/dev_efuse.h b/arch/fs1/arc/dev_efuse.h
index e16d951..6a281e2 100644
--- a/arch/fs1/arc/dev_efuse.h
+++ b/arch/fs1/arc/dev_efuse.h
@@ -6,6 +6,8 @@
 #ifndef _DEV_EFUSE_H_
 #define _DEV_EFUSE_H_
 
+#include <fs1_efuse.h>
+
 #define FLASH_VALID_MASK			0x1FFFFF
 
 #define FUSE_READ_FSM_TRIGGER		0x1
diff --git a/common/common.c b/common/common.c
index 16484ac..5c54d09 100644
--- a/common/common.c
+++ b/common/common.c
@@ -9,6 +9,8 @@
 #include <common.h>
 #include <delay.h>
 #include <spi_flash.h>
+#include <efuse.h>
+#include <dev_efuse.h>
 
 void boot_rom_init(void)
 {
@@ -36,3 +38,26 @@
 
 	spi_flash_init();
 }
+
+void boot_rom_fuse(void)
+{
+	if (boot_rom_fuse_init() != BOOT_ROM_SUCCESS)
+		goto efuse_fail_halt;
+
+	if (boot_rom_read_config_fuse() != BOOT_ROM_SUCCESS)
+		goto efuse_fail_halt;
+
+	if (!fuse_valid_volt_prot_pattern(boot_rom_global.config_fuse))
+		goto efuse_fail_halt;
+
+	boot_rom_set_efuse_read();
+
+	/* TODO: FS1 SKU decession pending */
+
+	/* TODO: Enable FAKE FUSE & RMA support */
+	return;
+
+efuse_fail_halt:
+	boot_rom_status_update(PROGRESS_INIT_EFUSE_INIT_FAIL);
+	boot_rom_halt();
+}
diff --git a/common/common.h b/common/common.h
index b0cf8ca..1472203 100644
--- a/common/common.h
+++ b/common/common.h
@@ -8,10 +8,14 @@
 
 #include <stdio.h>
 #include <stdint.h>
+#include <boot_rom.h>
 
 void boot_rom_clear_status(void);
 void boot_rom_persistent_dccm_init(uint8_t die_id);
 void boot_rom_read_cold_flops(void);
 void boot_rom_iccm_init(void);
+void boot_rom_halt(void);
+enum boot_rom_status boot_rom_read_config_fuse(void);
+void boot_rom_set_efuse_read(void);
 
 #endif /* _COMMON_H_ */
diff --git a/common/efuse.h b/common/efuse.h
index f3bbe77..45e8ff3 100644
--- a/common/efuse.h
+++ b/common/efuse.h
@@ -9,7 +9,6 @@
 #include <stdint.h>
 #include <dev_efuse.h>
 #include <spi_flash.h>
-#include <fs1_efuse.h>
 
 uint32_t boot_rom_fuse_read(uint32_t word, uint32_t *val);
 uint32_t boot_rom_fuse_init(void);
diff --git a/common/spi_flash.c b/common/spi_flash.c
index 9d6e0d3..b7f1c20 100644
--- a/common/spi_flash.c
+++ b/common/spi_flash.c
@@ -4,10 +4,10 @@
  * All Rights Reserved.
  */
 
-#include <fs1_efuse.h>
 #include <boot_rom.h>
 #include <spi_flash.h>
 #include "dev_spi_flash.h"
+#include <dev_efuse.h>
 
 void spi_flash_init(void)
 {
